File: 백준/Silver/9012. 괄호/괄호.py
n = int(input())
for i in range(n):
    bool_val = False
    arr = list(str(input()))
    stack = []
    for j in range(len(arr)):
        if arr[j] == '(':
            stack.append('(')
        else:
            if len(stack)>0:
                stack.pop()
            elif len(stack)==0:
                if arr[j] == ')':
                    bool_val = True
                    break;
    if bool_val == True:
        print('NO')
    elif len(stack)==0:
        print('YES')
    else:
        print('NO')

File: 백준/Silver/9012. 괄호/괄호.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());

        for (int i = 0; i < n; i++) {
            boolean flag = false;
            Stack<String> stack = new Stack<>();
            String[] arr = br.readLine().split("");
            for(String j:arr){
                if(j.equals("(")) stack.push("(");
                else if (j.equals(")")) {
                    if(stack.size()==0){
                        flag = true;
                        break;
                    }
                    stack.pop();
                }
            }
            if(stack.size()!=0||flag) System.out.println("NO");
            else System.out.println("YES");
        }
    }
}